### PR TITLE
Feature - add Azure Functions Databricks project template

### LINKED ISSUE #246: Provide template for an Azure Function that reports Databrick metrics on a scheduled time
**Is your feature request related to a problem? Please describe.**
Provide template for an Azure Function that reports Databrick metrics on a scheduled time

### PR DESCRIPTION
**Pull request by stijnmoreels**
Adds a project template which is an Azure Function with a timer that reports on each cycle the finished Databricks job runs as metrics.

- [x] Project template
- [x] Preview feature doc update
- [x] Integration test that runs the project

Closes #246

### CODE REVIEW

**Comment by stijnmoreels** [comment]
@tomkerkhove , first rundown of the new template. I'll check next time for an integration test that we'll eventually will set on `[Skip]` but just to make sure that the project runs correctly.

**Comment by stijnmoreels** [comment]
Getting there for the integration tests, now working on syncing all the components together and updating the YAML so we can run it on our remote CI build machine.

File: src/Arcus.Templates.AzureFunctions.Databricks/Arcus.Templates.AzureFunctions.Databricks.csproj [hunk 1]
```diff
@@ -0,0 +1,66 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+    <AzureFunctionsVersion>v3</AzureFunctionsVersion>
+    <DockerDefaultTargetOS>Linux</DockerDefaultTargetOS>
+    <!--#if (AuthoringMode)-->
+    <Authors>Arcus</Authors>
+    <Company>Arcus</Company>
+    <RepositoryType>Git</RepositoryType>
+    <Description>Provide a template to easily build Azure Functions running in Azure that report finished Databricks job runs as metrics.</Description>
```
Comment (tomkerkhove) [inline src/Arcus.Templates.AzureFunctions.Databricks/Arcus.Templates.AzureFunctions.Databricks.csproj:10]:
```suggestion
    <Description>Provide a template to easily build an Azure Functions that reports the completed Databricks job runs as metrics.</Description>
```

File: src/Arcus.Templates.AzureFunctions.Databricks/Arcus.Templates.AzureFunctions.Databricks.csproj [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    <IsPackable>true</IsPackable>
+    <PackageId>Arcus.Templates.AzureFunctions.Databricks</PackageId>
+    <Title>Template for Azure Functions Databricks project</Title>
+    <PackageType>Template</PackageType>
+    <PackageTags>Azure;Azure Functions;App Services;Databricks;Metric</PackageTags>
```
Comment (tomkerkhove) [inline src/Arcus.Templates.AzureFunctions.Databricks/Arcus.Templates.AzureFunctions.Databricks.csproj:15]:
```suggestion
    <PackageTags>Azure;Azure Functions;Databricks;Metric</PackageTags>
```

File: src/Arcus.Templates.AzureFunctions.Databricks/DatabricksMetricFunction.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+        /// <summary>
+        /// Reports the finished Databricks job runs as metric.
+        /// </summary>
+        /// <param name="timer">The timer instance to provide information about this current scheduled run.</param>
+        /// <param name="logger">The logger instance to write diagnostic trace messages and reporting metrics during the Databricks interaction.</param>
+        [FunctionName("DatabricksMetricFunction")]
+        public async Task Run([TimerTrigger("0 */1 * * * *")]TimerInfo timer, ILogger logger)
+        {
+            logger.LogInformation($"C# Timer trigger function executed at: {DateTime.Now}");
```
Comment (tomkerkhove) [inline src/Arcus.Templates.AzureFunctions.Databricks/DatabricksMetricFunction.cs:44]:
```suggestion
            logger.LogInformation($"C# Timer trigger function executed at: {DateTime.UtcNow}");
```

File: src/Arcus.Templates.AzureFunctions.Databricks/DatabricksMetricFunction.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            _configuration = configuration;
+            _secretProvider = secretProvider;
+        }
+
+        /// <summary>
+        /// Reports the finished Databricks job runs as metric.
+        /// </summary>
+        /// <param name="timer">The timer instance to provide information about this current scheduled run.</param>
+        /// <param name="logger">The logger instance to write diagnostic trace messages and reporting metrics during the Databricks interaction.</param>
+        [FunctionName("DatabricksMetricFunction")]
```
Comment (tomkerkhove) [inline src/Arcus.Templates.AzureFunctions.Databricks/DatabricksMetricFunction.cs:41]:
```suggestion
        [FunctionName("databricks-job-metrics")]
```

File: src/Arcus.Templates.AzureFunctions.Databricks/DatabricksMetricFunction.cs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+using Microsoft.Azure.WebJobs;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+
+namespace Arcus.Templates.AzureFunctions.Databricks
+{
+    /// <summary>
+    /// Represents an Azure Function that will report periodically the finished Databricks job runs as metrics.
+    /// </summary>
+    public class DatabricksMetricFunction
```
Comment (tomkerkhove) [inline src/Arcus.Templates.AzureFunctions.Databricks/DatabricksMetricFunction.cs:16]:
```suggestion
    public class DatabricksJobMetricsFunction
```

File: src/Arcus.Templates.AzureFunctions.Databricks/DatabricksMetricFunction.cs [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+            logger.LogInformation($"C# Timer trigger function executed at: {DateTime.Now}");
+
+            var metricName = _configuration.GetValue<string>("Arcus:ApplicationInsights:MetricName");
+            var baseUrl = _configuration.GetValue<string>("Arcus:Databricks:Url");
+            string secretToken = await _secretProvider.GetRawSecretAsync("Arcus.Databricks.SecretToken");
+
+            var startOfWindow = timer.ScheduleStatus.Last;
+            var endOfWindow = timer.ScheduleStatus.Next;
+
+            using var client = DatabricksClient.CreateClient(baseUrl, secretToken);
```
Comment (tomkerkhove) [inline src/Arcus.Templates.AzureFunctions.Databricks/DatabricksMetricFunction.cs:53]: Can't we use DI for this?

https://docs.microsoft.com/en-us/azure/azure-functions/functions-dotnet-dependency-injection

File: src/Arcus.Templates.AzureFunctions.Databricks/Startup.cs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+//[#if DEBUG]
+                stores.AddConfiguration(config);
+//[#endif]
+
+                //#error Please provide a valid secret provider, for example Azure Key Vault: https://security.arcus-azure.net/features/secrets/consume-from-key-vault
+                stores.AddAzureKeyVaultWithManagedServiceIdentity("https://your-keyvault-vault.azure.net/");
+            });
+
+            var instrumentationKey = config.GetValue<string>("Arcus:ApplicationInsights:InstrumentationKey");
+            var configuration = new LoggerConfiguration()
```
Comment (tomkerkhove) [inline src/Arcus.Templates.AzureFunctions.Databricks/Startup.cs:39]: Careful here, this is an azure function

https://observability.arcus-azure.net/guidance/use-with-dotnet-and-functions.html

File: docs/preview/index.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+## Azure Service Bus Topic
+
+```shell
+PM > dotnet new --install Arcus.Templates.ServiceBus.Topic
+```
+
+Read [here](features/servicebus-topic-worker-template) for standard and configurable features.
+
+## Azure Functions Databricks
```
Comment (tomkerkhove) [inline docs/preview/index.md:35]:
```suggestion
## Databricks Job Metrics (Azure Functions)
```

File: docs/preview/index.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+```shell
+PM > dotnet new --install Arcus.Templates.AzureFunctions.Databricks
```
Comment (tomkerkhove) [inline docs/preview/index.md:38]:
```suggestion
PM > dotnet new --install Arcus.Templates.AzureFunctions.Databricks.JobMetrics
```

File: docs/preview/features/azurefunctions-databricks-template.md [hunk 10]
```diff
@@ -0,0 +1,26 @@
+---
+title: "Azure Functions Databricks template"
```
Comment (tomkerkhove) [inline docs/preview/features/azurefunctions-databricks-template.md:2]:
```suggestion
title: "Databricks Job Metrics template (Azure Functions)"
```

File: docs/preview/features/azurefunctions-databricks-template.md [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+layout: default
+---
+
+# Azure Functions Databricks project template
```
Comment (tomkerkhove) [inline docs/preview/features/azurefunctions-databricks-template.md:6]:
```suggestion
# Databricks Job Metrics template (Azure Functions)
```

**Review by tomkerkhove — COMMENTED** [review]
Did a bunch of suggestions, feel free to let me know what you think!

File: docs/preview/index.md (hunk 9 above)
Comment (stijnmoreels) [reply docs/preview/index.md:38]: So you want another namespace and project alias too?

File: src/Arcus.Templates.AzureFunctions.Databricks/Startup.cs (hunk 7 above)
Comment (stijnmoreels) [reply src/Arcus.Templates.AzureFunctions.Databricks/Startup.cs:39]: 😅 💦

File: src/Arcus.Templates.AzureFunctions.Databricks/DatabricksMetricFunction.cs (hunk 6 above)
Comment (stijnmoreels) [reply src/Arcus.Templates.AzureFunctions.Databricks/DatabricksMetricFunction.cs:53]: Hmm, it depends on how big the `DatabricksClient` instance is to hold available during the entire life time of the Azure Function or if it's better to close the client after each run. If the instance is heavy to hold, then it would be rather extra 'big' memory consumption when the Azure Function isn't even running, right?

File: src/Arcus.Templates.AzureFunctions.Databricks/DatabricksMetricFunction.cs (hunk 5 above)
Comment (stijnmoreels) [reply src/Arcus.Templates.AzureFunctions.Databricks/DatabricksMetricFunction.cs:16]: Done, resolving.

File: src/Arcus.Templates.AzureFunctions.Databricks/DatabricksMetricFunction.cs (hunk 4 above)
Comment (stijnmoreels) [reply src/Arcus.Templates.AzureFunctions.Databricks/DatabricksMetricFunction.cs:41]: Done, resolving.

File: src/Arcus.Templates.AzureFunctions.Databricks/Startup.cs (hunk 7 above)
Comment (stijnmoreels) [reply src/Arcus.Templates.AzureFunctions.Databricks/Startup.cs:39]: Updated with sample and custom application component name.

File: docs/preview/index.md (hunk 9 above)
Comment (tomkerkhove) [reply docs/preview/index.md:38]: I'm just thinking long-term in case we provide other templates but if it's too much work then it's ok

Comment (stijnmoreels) [reply docs/preview/index.md:38]: It's some work, but can be done.

Comment (stijnmoreels) [reply docs/preview/index.md:38]: Done, resolving.

File: src/Arcus.Templates.AzureFunctions.Databricks.JobMetrics/Startup.cs [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+                //#error Please provide a valid secret provider, for example Azure Key Vault: https://security.arcus-azure.net/features/secrets/consume-from-key-vault
+                stores.AddAzureKeyVaultWithManagedServiceIdentity("https://your-keyvault-vault.azure.net/");
+            });
+
+            var instrumentationKey = config.GetValue<string>("Arcus:ApplicationInsights:InstrumentationKey");
+            var configuration = new LoggerConfiguration()
+                .MinimumLevel.Debug()
+                .MinimumLevel.Override("Microsoft", LogEventLevel.Information)
+                .Enrich.FromLogContext()
+                .Enrich.WithComponentName("Docker Hub Job Metrics")
```
Comment (tomkerkhove) [inline src/Arcus.Templates.AzureFunctions.Databricks.JobMetrics/Startup.cs:45]: This is not related to Docker Hub:
```suggestion
                .Enrich.WithComponentName("Azure Databricks Metrics Scraper")
```

File: src/Arcus.Templates.AzureFunctions.Databricks.JobMetrics/Startup.cs [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+            IConfiguration config = builder.GetContext().Configuration;
+
+            builder.ConfigureSecretStore(stores =>
+            {
+//[#if DEBUG]
+                stores.AddConfiguration(config);
+//[#endif]
+
+                //#error Please provide a valid secret provider, for example Azure Key Vault: https://security.arcus-azure.net/features/secrets/consume-from-key-vault
+                stores.AddAzureKeyVaultWithManagedServiceIdentity("https://your-keyvault-vault.azure.net/");
```
Comment (tomkerkhove) [inline src/Arcus.Templates.AzureFunctions.Databricks.JobMetrics/Startup.cs:37]: Let's add enviromnent variables by default here as well

File: src/Arcus.Templates.AzureFunctions.Databricks.JobMetrics/Startup.cs [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+            });
+
+            var instrumentationKey = config.GetValue<string>("Arcus:ApplicationInsights:InstrumentationKey");
```
Comment (tomkerkhove) [inline src/Arcus.Templates.AzureFunctions.Databricks.JobMetrics/Startup.cs:40]:
```suggestion
            var instrumentationKey = config.GetValue<string>("APPLICATIONINSIGHTS_INSTRUMENTATIONKEY");
```

**Review by tomkerkhove — APPROVED** [review]
LGTM, just a few more tweaks

File: src/Arcus.Templates.AzureFunctions.Databricks.JobMetrics/Startup.cs (hunk 14 above)
Comment (stijnmoreels) [reply src/Arcus.Templates.AzureFunctions.Databricks.JobMetrics/Startup.cs]: Ok, I'll change this and the one in the tests where it's being set. Resolving.

File: src/Arcus.Templates.AzureFunctions.Databricks.JobMetrics/Startup.cs (hunk 13 above)
Comment (stijnmoreels) [reply src/Arcus.Templates.AzureFunctions.Databricks.JobMetrics/Startup.cs:39]: Yes, ok, resolving.
